File: Naive/src/Grafo.java
import java.text.DecimalFormat;
import java.util.*;

public class Grafo {

    private Integer base;
    private Integer antibase;
    private int numVertices;
    private Map<Integer, List<Aresta>> vertices;
    public Grafo(int numVertices){
        this.numVertices = numVertices;
        this.vertices = new HashMap<Integer, List<Aresta>>();

    }

    public Integer getBase() {
        return base;
    }

    public Integer getAntibase() {
        return antibase;
    }

    public void addVertice(Vertice vertice){
        vertices.put(vertice.getId(), new ArrayList<>());
    }

    public void addAresta(Vertice verticeOrigem, Vertice verticeDestino,double peso) {
        List<Aresta> arestas = this.vertices.get(verticeOrigem.getId());
        if (arestas == null) {
            arestas = new ArrayList<>();
            this.vertices.put(verticeOrigem.getId(), arestas);
        }
        arestas.add(new Aresta(verticeOrigem, verticeDestino,peso));
    }



    public static Grafo gerarGrafoAleatorio(int nVertices, double probabilidade) {
        Grafo grafo = new Grafo(nVertices);

        for (int i = 0; i < nVertices; i++) {
            grafo.addVertice(new Vertice(i));
        }

        // Adicionar arestas aleatoriamente
        for (int i = 0; i < nVertices; i++) {
            for (int j = i + 1; j < nVertices; j++) {
                if(Math.random() < probabilidade){
                    double peso = Math.floor(Math.random() * (100 - 1 + 1)) + 1;
                    grafo.addAresta(new Vertice(i), new Vertice(j),peso);
                }
            }
        }

        return grafo;
    }

    public void printGrafo(){
        for (Integer origem : vertices.keySet()){
            List<Aresta> arestas = vertices.get(origem);

            for (Aresta aresta : arestas){
                DecimalFormat df = new DecimalFormat("#.##");
                System.out.println(origem + " -----"+ df.format(aresta.getPeso())  + " -----> " + aresta.getDestino().getId() );

            }
        }
    }

    public List<Integer> bfs(Vertice origem) {
        List<Integer> visitados = new ArrayList<>();
        Queue<Vertice> fila = new LinkedList<>();

        fila.add(origem);
        visitados.add(origem.getId());

        while (!fila.isEmpty()) {
            Vertice atual = fila.poll();
            List<Aresta> vizinhos = this.vertices.get(atual.getId());
            if (vizinhos == null) {
                continue;
            }
            for (Aresta adjacente : vizinhos) {
                int idAdjacente = adjacente.destino.getId();
                if (!visitados.contains(idAdjacente)) {
                    fila.add(adjacente.destino);
                    visitados.add(idAdjacente);
                }
            }

        }
        return visitados;
    }


    public List<Integer> encontrarBase() {
        int maiorFechoTransitivo = -1;
        List<Integer> bases = new ArrayList<>();

        for (int i = 0; i < numVertices; i++) {
            List<Integer> fechoTransitivo = bfs(new Vertice(i));
            int tamanhoFechoTransitivo = fechoTransitivo.size();
            if (tamanhoFechoTransitivo > maiorFechoTransitivo) {
                bases.clear();
                bases.add(i);
                maiorFechoTransitivo = tamanhoFechoTransitivo;
            } else if (tamanhoFechoTransitivo == maiorFechoTransitivo) {
                bases.add(i);
            }
        }

        return bases;
    }


    public List<Integer> encontrarAntiBase() {
        int menorFechoTransitivo = Integer.MAX_VALUE;
        int indiceBase = -1;
        List<Integer> antibase = new ArrayList<>();

        for (int i = 0; i < numVertices; i++) {
            List<Integer> fechoTransitivo = bfs(new Vertice(i));
            int tamanhoFechoTransitivo = fechoTransitivo.size();

            if (tamanhoFechoTransitivo < menorFechoTransitivo) {
                menorFechoTransitivo = tamanhoFechoTransitivo;
                indiceBase = i;
                antibase.clear();
                antibase.add(i);
            } else if (tamanhoFechoTransitivo == 1) {
                antibase.add(i);
            }
        }

        return antibase;
    }





    public List<Integer> transitivoInverso(Vertice origem) {
        List<Integer> visitados = new ArrayList<>();
        Stack<Vertice> pilha = new Stack<>();
        Map<Integer, List<Aresta>> arestasInvertidas = inverterArestas();

        pilha.add(origem);
        visitados.add(origem.getId());

        while (!pilha.isEmpty()) {
            Vertice atual = pilha.pop();
            List<Aresta> vizinhos = arestasInvertidas.get(atual.getId());
            if (vizinhos == null) {
                continue;
            }
            for (Aresta adjacente : vizinhos) {
                int idAdjacente = adjacente.destino.getId();
                if (!visitados.contains(idAdjacente)) {
                    pilha.push(adjacente.destino);
                    visitados.add(idAdjacente);
                }
            }
        }

        return visitados;
    }

    private Map<Integer, List<Aresta>> inverterArestas() {
        Map<Integer, List<Aresta>> arestasInvertidas = new HashMap<>();
        for (Integer origem : vertices.keySet()) {
            List<Aresta> arestas = vertices.get(origem);
            for (Aresta aresta : arestas) {
                int destino = aresta.destino.getId();
                List<Aresta> arestasDestino = arestasInvertidas.get(destino);
                if (arestasDestino == null) {
                    arestasDestino = new ArrayList<>();
                    arestasInvertidas.put(destino, arestasDestino);
                }
                arestasDestino.add(new Aresta(aresta.destino, new Vertice(origem), aresta.getPeso()));
            }
        }
        return arestasInvertidas;
    }




}

File: src/Grafo.ts
import Vertice from "./Vertice";

export default class Grafo {
    numVertices: number;
    // vertices: Map<number, Array<Aresta>>
    vertices: Vertice[];

    constructor(numVertices: number){
        this.numVertices = numVertices;
        // this.vertices = new Map<number, Array<Aresta>>();
        this.vertices = [];
    }
 
    
    localizarVertice(id: number): Vertice{
        return this.vertices[id];
    }
 

    /**
     * Adiciona um vértice com sua respectiva lista de vértices adjacentes ao grafo
     * @param {number} vertice 
     */
    addVertece(vertice: number): void{
        this.vertices.push(new Vertice(vertice));
    }

    /**
     * Adiciona uma aresta entre um vértice de origem e um vértice de destino
     * @param {number} verticeOrigem 
     * @param {number} verticeDestino
     * @param {number} peso peso da Aresta
     */
    addAresta(verticeOrigem: number, verticeDestino: number, peso: number): void{
        const origem = this.localizarVertice(verticeOrigem);
        const destino = this.localizarVertice(verticeDestino);

        origem.addAresta(origem, destino);
      
    
        //descomentar para que o grafo seja não-direcional
        //Fará com que toda aresta seja craida para as duas direções entre um vértice
        //this.vertices.get(w).push(new Aresta(v, w, peso));
    }

    /**
     * Função estática para criar grafos aleatórios baseado no algoritmo de Erdos Reyni
     * @param {number} nVertices exata quantidade de vertices no grafo
     * @param {number} probabilidade valor de 0 a 1 que representa a probabilidade de se criar uma aresta a cada par de vértice
     * @returns retorna um grafo
     */
    static gerarGrafoAleatorio(nVertices: number, probabilidade: number): Grafo {
        const grafo: Grafo = new Grafo(nVertices);

        for (let i = 0; i < nVertices; i++) {
            grafo.addVertece(i);
        }
      
        // Adicionar arestas aleatoriamente
        for (let i = 0; i < nVertices; i++) {
            for (let j = i + 1; j < nVertices; j++) {
                if(Math.random() < probabilidade) {
                    const peso = Math.floor(Math.random() * (100 - 1 + 1)) + 1;
                    grafo.addAresta(i, j, peso);
                }
            }
        }
      
        return grafo;
      }
      

    
    /**
     * Gera uma matriz a partir do objeto instanciado e dos vértices e arestas que foram adicionados a ele.
     * @returns uma matriz de adjacencia do grafo
     */
    gerarMatriz(): number[][]{
        const matriz: Array<Array<number>> = new Array(this.numVertices);
        for(let i = 0; i < this.numVertices; i++){
            matriz[i] = new Array(this.numVertices).fill(0);
        }

        for(let linha = 0; linha < this.numVertices; linha++){

            const adjacentes = this.vertices[linha].getAdjacentes;

            for(let coluna = 0; coluna < this.numVertices; coluna++){
                if(linha == coluna){
                    matriz[linha][coluna] = 0;
                }
                else{
                    for(const a of adjacentes){
                        if(a.getDestino.getId == coluna){
                            matriz[linha][coluna] = 1;
                            break;
                        }
                    }
                }
                
            }
        }

        return matriz;

    }

    /**
     * Algoritmo de busca em largura para buscar todos os vértices que são alcançaveis a partir de um vértice de origem 
     * @param origem vertice de origem para a busca em largura
     * @returns array de vértices que são alcançaveis pelo vértice de origem
     */
    bfs(origem: Vertice){
        const distancia: number[] = [];
        const arvore: Vertice[] | null[] = [];
        const matrizAdjacentes: number[] = [];
        const fila: Vertice[] = [];

        for(let i = 0; i < this.numVertices; i++){
            this.vertices[i].setVisitado = false;
            distancia[i] = Infinity;
            arvore[i] = null
            matrizAdjacentes[i] = 0;
        }

        this.vertices[origem.getId].visitar();
        distancia[origem.getId] = 0;
        fila.push(this.vertices[origem.getId]);

        while(fila.length > 0){
            const v = fila.shift();
            if(v){
                const vizinhos = v.verticesVizinhos();

                for(const vizinho of vizinhos){
                    if(!vizinho.getVisitado){
                        vizinho.visitar();
                        arvore[vizinho.getId] = vizinho;
                        matrizAdjacentes[vizinho.getId] = 1;
                        distancia[vizinho.getId] = distancia[v.getId] + 1;
                        fila.push(vizinho);
                    }
                }

            }
            
        }

        arvore[origem.getId] = origem;
        matrizAdjacentes[origem.getId] = 1;
        return matrizAdjacentes;
    }


    /**
     * Gera uma matriz de fechos transitivos a partir da busca em largura
     * @returns matriz de fechos transitivos do grafo
     */
    fechoTransitivoNaive(){
        const fechosTransitivos: Array<Array<number>>  = [];
        for(const vertice of this.vertices){
            fechosTransitivos.push(this.bfs(vertice));
        }

        return fechosTransitivos;
    }


    /**
     * ### Não recomendado para grafos acima de 3 mil vértices!
     * Método para gerar uma matriz de fechos transitivos diretos de um grafo qualquer utilizando o algoritmo de warshall. As linhas da matriz representam o fecho transitivo direto de um vértice e as colunas representam os fechos transitivos inversos
     * @param matriz um grafo em forma de matriz de adjacencia
     * @returns uma matriz que representa os fechos transitivos diretos que um grafo
     */
    fechoTransitivoWarshall(matriz: number[][]){
        for(let i = 0; i < this.numVertices; i++){
            matriz[i][i] = 1;
        }

        for(let k = 0; k < this.numVertices; k++){
            for(let i = 0; i < this.numVertices; i++){
                for(let j = 0; j < this.numVertices; j++){
                    if(matriz[i][j] == 0 && matriz[i][k] == 1 && matriz[k][j] == 1){
                        matriz[i][j] = 1;
                    }
                }
            }
        }

        return matriz;
    }


    /**
     * Esse método gera um array com os ids dos vértices que são a base de um grafo a partir de uma matriz de fecho transitivo do mesmo grafo
     * @param matrizFechosTransitivos uma matriz de adjacencia que represete os fechos transitivos de todos os pares de vértice de um grafo
     * @returns um array com o id de todos os vértices que pertencem a anti base 
     */
    gerarBase(matrizFechosTransitivos: number[][]){
        const base: number[] = [];

        for(let coluna = 0; coluna < this.numVertices; coluna++){
            let eBase = true;
           

            for(let linha = 0; linha < this.numVertices; linha++){
  
                if(linha != coluna && matrizFechosTransitivos[linha][coluna] == 1){
                    eBase = false;
                }
                
            }

            if(eBase){
                base.push(coluna);
            }
        }

        return base;

    }


    /**
     * Esse método gera um array com os ids dos vértices que são a anti base de um grafo a partir de uma matriz de fecho transitivo do mesmo grafo
     * @param matrizFechosTransitivos uma matriz de adjacencia que represete os fechos transitivos de todos os pares de vértice de um grafo
     * @returns um array com o id de todos os vértices que pertencem a anti base
     */
    gerarAntiBase(matrizFechosTransitivos: number[][]){
        const antiBase: number[] = [];

        for(let linha = 0; linha < this.numVertices; linha++){
            let eAntiBase = true;
           

            for(let coluna = 0; coluna < this.numVertices; coluna++){
  
                if(linha != coluna && matrizFechosTransitivos[linha][coluna] == 1){
                    eAntiBase = false;
                }
                
            }

            if(eAntiBase){
                antiBase.push(linha);
            }
        }

        return antiBase;  
    }




    /**
     * Concatena cada vértice do grafo apontando para todos os vértices que são adjacentes a ele em uma string
     */
    toString(): string{
        let print: string = "";

        for(let i = 0; i < this.vertices.length; i++){
            const verticesAdjacentes = this.vertices[i].getAdjacentes;

            for(const aresta of verticesAdjacentes){
                print += `${aresta.getOrigem} ---${aresta.getPeso}---> ${aresta.getDestino}\n`;
            }

        }
        
        return print;
    }
    
}
